--- openssl-1.1.1n/apps/speed.c	2022-03-15 22:37:47.000000000 +0800
+++ openssl-1.1.1n-new/apps/speed.c	2024-12-26 11:20:20.973830681 +0800
@@ -2635,15 +2635,24 @@
                         BIO_printf(bio_err, "\nEVP_CIPHER_CTX_new failure\n");
                         exit(1);
                     }
+
+#if 1
+                    if (decrypt)
+                        EVP_DecryptInit_ex(loopargs[k].ctx, evp_cipher, NULL,
+                                           key32, iv);
+                    else
+                        EVP_EncryptInit_ex(loopargs[k].ctx, evp_cipher, NULL,
+                                           key32, iv);
+#else
                     if (!EVP_CipherInit_ex(loopargs[k].ctx, evp_cipher, NULL,
                                            NULL, iv, decrypt ? 0 : 1)) {
                         BIO_printf(bio_err, "\nEVP_CipherInit_ex failure\n");
                         ERR_print_errors(bio_err);
                         exit(1);
                     }
-
+#endif
                     EVP_CIPHER_CTX_set_padding(loopargs[k].ctx, 0);
-
+#if 0
                     keylen = EVP_CIPHER_CTX_key_length(loopargs[k].ctx);
                     loopargs[k].key = app_malloc(keylen, "evp_cipher key");
                     EVP_CIPHER_CTX_rand_key(loopargs[k].ctx, loopargs[k].key);
@@ -2654,6 +2663,7 @@
                         exit(1);
                     }
                     OPENSSL_clear_free(loopargs[k].key, keylen);
+#endif
                 }
 
                 Time_F(START);
@@ -2677,11 +2687,11 @@
             }
         }
     }
-
+#if 0
     for (i = 0; i < loopargs_len; i++)
         if (RAND_bytes(loopargs[i].buf, 36) <= 0)
             goto end;
-
+#endif
 #ifndef OPENSSL_NO_RSA
     for (testnum = 0; testnum < RSA_NUM; testnum++) {
         int st = 0;
@@ -2774,11 +2784,11 @@
         }
     }
 #endif                          /* OPENSSL_NO_RSA */
-
+#if 0
     for (i = 0; i < loopargs_len; i++)
         if (RAND_bytes(loopargs[i].buf, 36) <= 0)
             goto end;
-
+#endif
 #ifndef OPENSSL_NO_DSA
     for (testnum = 0; testnum < DSA_NUM; testnum++) {
         int st = 0;

