diff --git a/include/tdep-riscv/libunwind_i.h b/include/tdep-riscv/libunwind_i.h
index 951de12..d4452cd 100644
--- a/include/tdep-riscv/libunwind_i.h
+++ b/include/tdep-riscv/libunwind_i.h
@@ -174,12 +174,8 @@ dwarf_getfp (struct dwarf_cursor *c, dwarf_loc_t loc, unw_fpreg_t *val)
 
   /* FIXME: unw_word_t may not be equal to FLEN */
   addr = DWARF_GET_LOC (loc);
-#if __riscv_xlen == __riscv_flen
   return (*c->as->acc.access_mem) (c->as, addr, (unw_word_t *) valp,
                                        0, c->as_arg);
-#else
-# error "FIXME"
-#endif
 }
 
 static inline int
@@ -197,12 +193,8 @@ dwarf_putfp (struct dwarf_cursor *c, dwarf_loc_t loc, unw_fpreg_t val)
 
   /* FIXME: unw_word_t may not be equal to FLEN */
   addr = DWARF_GET_LOC (loc);
-#if __riscv_xlen == __riscv_flen
   return (*c->as->acc.access_mem) (c->as, addr, (unw_word_t *) valp,
                                        1, c->as_arg);
-#else
-# error "FIXME"
-#endif
 }
 
 static inline int
