From 53e4d5e18ebf11d6d48868101728ec62f86af6d8 Mon Sep 17 00:00:00 2001
From: anruliu <anruliu@allwinnertech.com>
Date: Wed, 4 Jan 2023 19:17:12 +0800
Subject: [PATCH] fix compilation errors

Signed-off-by: anruliu <anruliu@allwinnertech.com>
---
 include/ion/ion.h | 8 ++++++--
 ion.c             | 9 +++++++--
 2 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/include/ion/ion.h b/include/ion/ion.h
index 1480bd9..eb80cf1 100644
--- a/include/ion/ion.h
+++ b/include/ion/ion.h
@@ -24,7 +24,9 @@
 #include <sys/types.h>
 #include <linux/ion.h>
 
-__BEGIN_DECLS
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 struct ion_handle;
 
@@ -51,6 +53,8 @@ int ion_query_get_heaps(int fd, int cnt, void* buffers);
 int ion_is_legacy(int fd);
 int ion_is_using_modular_heaps(int fd);
 
-__END_DECLS
+#ifdef __cplusplus
+}
+#endif
 
 #endif /* __SYS_CORE_ION_H */
diff --git a/ion.c b/ion.c
index cce583a..118c1a2 100644
--- a/ion.c
+++ b/ion.c
@@ -22,7 +22,7 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <linux/ion.h>
-#include <stdatomic.h>
+//#include <stdatomic.h>
 #include <stdio.h>
 #include <string.h>
 #include <sys/ioctl.h>
@@ -33,7 +33,12 @@
 #include <ion/ion.h>
 #include <linux/ion_4.19.h>
 
-#include <log/log.h>
+//#include <log/log.h>
+#include <atomic>
+using namespace std;
+
+//#include <log/log.h>
+#define ALOGE(format,...)  printf("\n[Log Error] [File:%s, Line:%d] " format "\n", __FILE__, __LINE__, ##__VA_ARGS__)
 
 #define ION_ABI_VERSION_MODULAR_HEAPS 2
 
-- 
2.25.1

