diff --git a/CMakeLists.txt b/CMakeLists.txt
index dfbd0c5..3284a3e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -83,6 +83,7 @@ option(EXPAT_BUILD_TOOLS "build the xmlwf tool for expat library" ${_EXPAT_BUILD
 option(EXPAT_BUILD_EXAMPLES "build the examples for expat library" ON)
 option(EXPAT_BUILD_TESTS "build the tests for expat library" ON)
 option(EXPAT_SHARED_LIBS "build a shared expat library" ON)
+option(EXPAT_STATIC_LIBS "build a static expat library" ON)
 option(EXPAT_BUILD_DOCS "build man page for xmlwf" ${_EXPAT_BUILD_DOCS_DEFAULT})
 option(EXPAT_BUILD_FUZZERS "build fuzzers for the expat library" OFF)
 option(EXPAT_BUILD_PKGCONFIG "build pkg-config file" ${_EXPAT_BUILD_PKGCONFIG_DEFAULT})
@@ -311,6 +312,9 @@ else()
     set(_SHARED STATIC)
 endif()
 
+if(EXPAT_STATIC_LIBS)
+    set(_STATIC STATIC)
+endif()
 # Avoid colliding with Expat.dll of Perl's XML::Parser::Expat
 if(WIN32 AND NOT MINGW)
     set(_EXPAT_OUTPUT_NAME libexpat)  # CMAKE_*_POSTFIX applies, see above
@@ -323,6 +327,9 @@ else()
 endif()
 
 add_library(expat ${_SHARED} ${expat_SRCS})
+add_library(expat_static ${_STATIC} ${expat_SRCS})
+set_target_properties(expat_static PROPERTIES OUTPUT_NAME "expat")
+
 if(EXPAT_WITH_LIBBSD)
     target_link_libraries(expat ${LIB_BSD})
 endif()
