From abc60e7d5333652bd5466529913df7b70bf45fa3 Mon Sep 17 00:00:00 2001
From: luoyuyi <luoyuyi@allwinnertech.com>
Date: Thu, 14 Sep 2023 10:36:21 +0800
Subject: [PATCH] aic8800: bt mac preferentially use addr_mgt

---
 tools/hciattach_aic.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/tools/hciattach_aic.c b/tools/hciattach_aic.c
index a5ae5a9..efc3756 100755
--- a/tools/hciattach_aic.c
+++ b/tools/hciattach_aic.c
@@ -105,6 +105,10 @@
 #define AIC_BT_CONF_PATH_NAME "/etc/bluetooth/aic_bt.conf"
 #endif
 
+#ifndef ADDR_MGT_BT_PATH_NAME
+#define ADDR_MGT_BT_PATH_NAME "/sys/class/addr_mgt/addr_bt"
+#endif
+
 typedef void (*hci_cback)(void *);
 
 typedef struct {
@@ -326,10 +330,26 @@ static int generate_bdaddr(unsigned char *buf)
 	int fd;
 	FILE* conf_fd = NULL;
 	unsigned mac_hex;
-	int i = 12; /* from MSB to LSB*/
+	int i; /* from MSB to LSB*/
+	int addr_bt_fd;
+	bdaddr_t bdaddr;
+	char addr_bt[18] = {'\0'};
+
+	addr_bt_fd = open(ADDR_MGT_BT_PATH_NAME, O_RDWR);
+	if (addr_bt_fd > 0) {
+		read(addr_bt_fd, addr_bt, sizeof(addr_bt) - 1);
+		str2ba(addr_bt, &bdaddr);
+		for (i = 12; i >= 7; i--) {
+			buf[i] = bdaddr.b[i-7];
+		}
+		close(addr_bt_fd);
+		if (check_bdaddr_valid(buf))
+			return 0;
+	}
 
 	conf_fd = fopen(AIC_BT_CONF_PATH_NAME, "r");
 	if(conf_fd) {
+		i = 12;
 		fscanf(conf_fd, "%x", &mac_hex);
 		while (!feof(conf_fd)) {
 			buf[i--] = mac_hex;
-- 
2.29.0

