From 410e56125ed590d05637ae1eb7d1228744debdb9 Mon Sep 17 00:00:00 2001
From: luoyuyi <luoyuyi@allwinnertech.com>
Date: Mon, 23 Jun 2025 16:06:10 +0800
Subject: [PATCH] support configure hciconfig to reduce space

---
 tools/hciconfig.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/tools/hciconfig.c b/tools/hciconfig.c
index ddc17c4..029792b 100644
--- a/tools/hciconfig.c
+++ b/tools/hciconfig.c
@@ -83,7 +83,7 @@ static void print_dev_list(int ctl, int flags)
 
 	free(dl);
 }
-
+#ifndef HCICONFIG_SIMP
 static void print_pkt_type(struct hci_dev_info *di)
 {
 	char *str;
@@ -472,6 +472,7 @@ static void cmd_encrypt(int ctl, int hdev, char *opt)
 		exit(1);
 	}
 }
+#endif
 
 static void cmd_up(int ctl, int hdev, char *opt)
 {
@@ -508,7 +509,7 @@ static void cmd_reset(int ctl, int hdev, char *opt)
 	cmd_down(ctl, hdev, "down");
 	cmd_up(ctl, hdev, "up");
 }
-
+#ifndef HCICONFIG_SIMP
 static void cmd_ptype(int ctl, int hdev, char *opt)
 {
 	struct hci_dev_req dr;
@@ -1863,6 +1864,7 @@ static void cmd_unblock(int ctl, int hdev, char *opt)
 
 	hci_close_dev(dd);
 }
+#endif
 
 static void print_dev_hdr(struct hci_dev_info *di)
 {
@@ -1900,6 +1902,7 @@ static void print_dev_info(int ctl, struct hci_dev_info *di)
 	printf("\tTX bytes:%d acl:%d sco:%d commands:%d errors:%d\n",
 		st->byte_tx, st->acl_tx, st->sco_tx, st->cmd_tx, st->err_tx);
 
+#ifndef HCICONFIG_SIMP
 	if (all && !hci_test_bit(HCI_RAW, &di->flags)) {
 		print_dev_features(di, 0);
 
@@ -1917,6 +1920,7 @@ static void print_dev_info(int ctl, struct hci_dev_info *di)
 		if (hci_test_bit(HCI_UP, &di->flags))
 			cmd_version(ctl, di->dev_id, NULL);
 	}
+#endif
 
 	printf("\n");
 }
@@ -1930,6 +1934,7 @@ static struct {
 	{ "up",		cmd_up,		0,		"Open and initialize HCI device" },
 	{ "down",	cmd_down,	0,		"Close HCI device" },
 	{ "reset",	cmd_reset,	0,		"Reset HCI device" },
+#ifndef HCICONFIG_SIMP
 	{ "rstat",	cmd_rstat,	0,		"Reset statistic counters" },
 	{ "auth",	cmd_auth,	0,		"Enable Authentication" },
 	{ "noauth",	cmd_auth,	0,		"Disable Authentication" },
@@ -1971,6 +1976,7 @@ static struct {
 		"\n\t\t\t3 - Non connectable undirected advertising"},
 	{ "noleadv",	cmd_no_le_adv,	0,		"Disable LE advertising" },
 	{ "lestates",	cmd_le_states,	0,		"Display the supported LE states" },
+#endif
 	{ NULL, NULL, 0 }
 };
 
-- 
2.25.1

